### ISSUE TITLE
S1694 should not raise when (the only) abstract member is internal

### ISSUE THREAD

**Issue by lg2de** [SonarSource/sonar-dotnet#3235]
### Description

Rule S1694: "An abstract class should have both abstract and concrete methods."

It should not apply if the (one and only) abstract method is internal.

The sample below wants to define a base class available public but with internal only unified interface.
An c# `interface` cannot define an internal member.
So the action `Convert this 'abstract' class to an interface.` is invalid.

Any recommendations?

### Repro steps

```cs
public abstract class BaseClass
{
    internal abstract void Serialize(IBufferWriter<byte> buffer);
}

public class DerivedClass : BaseClass
{
    public int Value { get; set; }

    internal static DerivedClass Deserialize(byte lengthSignature, ref SequenceReader<byte> reader)
    {
        return new DerivedClass { Value = reader.ReadI4(lengthSignature) };
    }

    internal override void Serialize(IBufferWriter<byte> buffer)
    {
        buffer.WriteI4(this.Value);
    }
}
```

### Expected behavior

The rule should not apply.

### Actual behavior

`warning S1694: Convert this 'abstract' class to an interface.`

### Known workarounds

* add dummy member
* disable rule

### Related information

* SonarC# Version: SonarAnalyzer.CSharp 8.4.0.15306
* Visual Studio Version: 16.4.5
* Operating System: Windows 10

**Comment by pavel-mikula-sonarsource** [comment]
Hi @lg2de,

The rule does what it's supposed to do. What you need is:

> The sample below wants to define a base class available public but with **internal only unified interface.**

To rephrase it, you need internal interface. There's no need for empty abstract class, just as S1694 suggests.

**Comment by lg2de** [comment]
> To rephrase it, you need internal interface. There's no need for empty abstract class, just as S1694 suggests.

If I define an internal interface I can provide public derived classes. But I do not have the public non-instantiable base class.

**Comment by pavel-mikula-sonarsource** [comment]
>But I do not have the public non-instantiable base class.

And what do you need it for? It's not instantiable. It does not contain any other methods with implementation. It can be used only as an interface. So it sounds like an interface.

This rule is supposed to warn you about this bad pattern.
